Привет! Меня зовут Исмаилов Исмаил, я буду проверять твой проект :) Можешь обращаться ко мне на «ты». Если тебе комфортно, то и я буду к тебе так обращаться, если нет, то обязательно скажи об этом.

Пожалуйста, не удаляй мои комментарии, которые я буду оставлять в работе. Это особенно поможет, если твои проекты будут отправлены на повторную проверку. При повторных проверках у комментариев будут приписки: "Вторая итерация", "Третья итерация" и т.д. 

Ты тоже можешь реагировать на мои комментарии, но в таком случае постарайся, чтобы твои комментарии отличались от моих: например, выделяй их своим любимым цветом — так у нас не возникнет путаницы :)

Кстати, про цвета! Мои комментарии будут в следующей цветовой гамме: зелеными, желтыми и красными. Например:

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
В случае, если решение на отдельном шаге является полностью правильным. </div>

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 
    
В случае, если решение может стать еще лучше с некоторыми корректировками </div>
 
<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 
    
В случае, если решение какого-то шага требует значительных изменений. Проект не может быть принят с первого раза, если ревью содержит комментарии, помеченные этим цветом </div>

<div class="alert alert-block alert-info">

Прощу прощения, я работал над проектом в другом месте, что и не тот отправил на проверку :(
    
</div>

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Вторая итерация </b> 

Не стоило удалять мои комментарии, не делай так больше

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Третья итерация </b> 

Напоминаю об этом: это даже не просьба, а правило Практикума. С удаленными комментариями работу не приму

<div class="alert alert-block alert-info">

Долго думал, что не так с комментариями, потом меня дошло, что нужно скорпировать комментарии с 1 итерации и вставить сюда. Я правильно сделал? Если не так, дай, пожалуйста, мне знать
    
</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [1]:
# Подключаем библиотеку pandas
import pandas as pd

# Получаем датасет
try:
    df = pd.read_csv('/datasets/users_behavior.csv')
except:
    df = pd.read_csv('users_behavior.csv')

# Смотрим и изучаем датасет
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [2]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [3]:
# Проверяем пропущенные значения
df.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

<div class="alert-success"> 
<b>Комментарий ревьюера 👍</b>

Хороший первичный обзор данных

Нет пропущенных значений, теперь переходим к следующему шагу

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 </b>

Не забывай о существовании текстовых ячеек

## Разбейте данные на выборки

In [4]:
# Подключаем библиотеку для деления на две выборки (Валидационную выборку)
from sklearn.model_selection import train_test_split

# Подключаем функцию для оценки правильности
from sklearn.metrics import accuracy_score

In [5]:
# Устанавливаем признаки и цели обучения
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

Нам нужно отделить по 20% данных для валидационной и тестовой выборки, а 60% - обучающей выборки

In [6]:
# Разделяем исходные данные на обучающую, валидационную и тестовую выборки

# Как обычно, будем разделять 60% данных для обучающей выборки, а features_temp и target_temp будут временным датафреймом
features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size= (1 - 0.6), random_state=12345) 

# А тестовая и валидационная - по 20% данных
features_valid, features_test, target_valid, target_test = train_test_split(features_temp, target_temp, test_size=(1 - 0.5), random_state=12345) 

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

У тебя нигде переменная model не задана, как можем мы ее использовать?

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Те же ошибки, model никак не инициализирована; мы ее ни на чем не обучаем, чтобы делать предсказания 

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌ Третья итерация </b>

Не создана тестовая выборка

<div class="alert alert-block alert-info">

~Обученная модель перемещена в 3 шаг - Исследовать модели~
    
</div>

<div class="alert alert-block alert-info">

Добавлена тестовая выборка
    
</div>

## Исследуйте модели

Нам требуется исследовать какими тарифом пользователь пользуются, а эти признаки являются категориальными, т.е. будем применять методом обучения для классификации

In [7]:
# Подключаем библиотеки для обучения разными способами:
# - Случайный лес в регрессии
# - Линейная регрессия
# - Дерево решений в регресии
# - Случайный лес
# - Решающее дерево

# from sklearn.ensemble import RandomForestRegressor
# from sklearn.linear_model import LinearRegression
# from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

# И библиотеку для вычисления MSE (Среднеквадратичная ошибка) - Для регрессии
# from sklearn.metrics import mean_squared_error

Исследуем качество модели разными способами

In [8]:
# Методом случайного леса

best_model = None
best_result = 0
for est in range(1, 11):
    # Обучим модель с заданным количеством деревьев
    model = RandomForestClassifier(random_state=12345, n_estimators=est) 
    
    # Обучим модель на тренировочной выборке
    model.fit(features_train, target_train) 
    
    # Посчитаем качество модели на валидационной выборке
    result = model.score(features_valid, target_valid) 
    
    if result > best_result:
        # Сохраним наилучшую модель
        best_model = model
        # Сохраним наилучшее значение метрики accuracy на валидационных данных
        best_result = result

In [9]:
print("Accuracy наилучшей модели на валидационной выборке:", best_result)

Accuracy наилучшей модели на валидационной выборке: 0.7853810264385692


In [10]:
# Методом решающего дерева

best_model = None
best_result = 0

for depth in range(1, 6):
    # Обучим модель с заданной глубиной дерева
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    
    # Обучим модель
    model.fit(features_train, target_train)
    
    # Получим предсказания модели
    predictions_valid = model.predict(features_valid)
    
    # Посчитаем качество модели
    result = accuracy_score(target_valid, predictions_valid) 
    
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth
    
    print("max_depth =", depth, ": ", end='')
    print(result)

max_depth = 1 : 0.7542768273716952
max_depth = 2 : 0.7822706065318819
max_depth = 3 : 0.7853810264385692
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7791601866251944


<div class="alert-danger"> 
<b>Комментарий ревьюера ❌ Третья итерация </b> 

Ты здесь обучаешься и делаешь предсказания на всей выборке (переменные features и target), так делать нельзя, для этого мы и разбивали выборку на три части

In [11]:
print("Accuracy лучшей модели:", best_result, '\nГлубина обучения:', best_depth)

Accuracy лучшей модели: 0.7853810264385692 
Глубина обучения: 3


<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Третья итерация </b> 

Ты сделал одну и ту же работу дважды, в прошлом пункте то же самое, только без вывода метрики. Почему сделал так?

<div class="alert alert-block alert-info">

Действительно так, я поторопился сдать проект до НГ, что и допустил много ошибок
    
</div>

Здесь видно, что максимальная глубина 3 является оптимальной по сравнению остальных 

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

RMSE - метрика для регрессии, а у нас задача классификации

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Опять же - у нас не регрессия, а классификация

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Почему где-то accuracy, где-то RMSE?

## Проверьте модель на тестовой выборке

In [12]:
# Обучим модель с заданным количеством деревьев
model = RandomForestClassifier(random_state=12345, n_estimators=est) 

# Обучим модель на тренировочной выборке
model.fit(features_train, target_train) 

# Получим предсказания обучающей модели
train_predictions = model.predict(features_train)

# Получим предсказания тестовой модели
test_predictions = model.predict(features_test)

In [13]:
print("Accuracy")
print("Обучающая выборка:", accuracy_score(target_train, train_predictions))
print("Тестовая выборка:", accuracy_score(target_test, test_predictions))

Accuracy
Обучающая выборка: 0.9823651452282157
Тестовая выборка: 0.7807153965785381


Здесь получилось, что обученная модель показалась себя намного лучше, ближе к 1.0, а тестовая справилась, но не намного хуже валидационного выборки, разница лишь в сотых, отн. погрешность около 0,6%

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡 Третья итерация </b>

Почему выбрал не ту модель, которая показала наилучшее качество ранее?

<div class="alert alert-block alert-info">

По той же причине - спешки сдать проект до НГ
    
</div>

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌ Третья итерация </b>

Ты не проверил модель на тестовой выборке, ты ее пока даже не создал

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Код не воспроизводится. Да и не нужно создавать дополнительную функцию, используй тот же accuracy

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Пункт не сделан :(

## (бонус) Проверьте модели на адекватность

In [14]:
# Подключаем библиотеку еще
from sklearn.dummy import DummyClassifier

In [15]:
# Устанавливаем стратегию прогноза как "самый частый случай", 
# т.е. частую метку класса в наблюдаемом аргументе target, 
# переданном для соответствия
dummy_clf = DummyClassifier(strategy = "most_frequent")

# Обучим модель
dummy_clf.fit(features_train, target_train)

# Создаем модель, указав стратению как "самый частый случай"
DummyClassifier(strategy = 'most_frequent')

# Получим предсказания модели
dummy_clf.predict(features_train)

# Посчитаем качество модели
dummy_clf.score(features_train, target_train)

0.6924273858921162

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 Третья итерация </b>

Можно сделать легче, посмотреть на баланс классов: например, если у нас 90 процентов таргета это "1", а остальные "10" - 0, то мы можем тупо выдавать каждый раз цифру 1 и получать метрику 90 %. Любая модель, которая будет выдавать качество ниже этого - плохая/неадекватная

Получилось, что обученная модель не является идеальным по сравнению остальных

In [16]:
df['is_ultra'].value_counts()

0    2229
1     985
Name: is_ultra, dtype: int64

<div class="alert alert-block alert-info">

Если присмотреться, то действительно да, тут можно посмотреть как 

2229 + 985 = 3214

100% - 3214

69% - 2229

31% - 985
    
</div>

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌ Третья итерация </b>

Нет верного разбиения на выборки, неверно обучается модель, неверно проверяется модель. Также оставил пару желтых комментарий, дай, пожалуйста, комментарий к ним 
    
И не удаляй комментарии.

<div class="alert-danger"> 
<b>Комментарий ревьюера ❌</b> 

Проект не доделан, код не воспроизводится. Возникло ощущение, что ты списывал и просто запутался в том, что делал. Подскажи, так ли это или нужна ли тебе какая-то помощь?

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
